vowpalwabbit/learner.h

```diff
@@ -122,23 +127,59 @@ inline float noop_sensitivity(void*, base_learner&, example&)
 }
 float recur_sensitivity(void*, base_learner&, example&);
 
+inline void increment_depth(example& ex)
+{
+  if(vw_dbg::track_stack)
+    ++ex.stack_depth;
+}
+
+inline void increment_depth(multi_ex& ec_seq)
+{
+  if(vw_dbg::track_stack){
+    for (auto& ec : ec_seq)
+    {
+      ++ec->stack_depth;
+    }
+  }
+}
+
+inline void decrement_depth(example& ex) {
+  if(vw_dbg::track_stack)
+    --ex.stack_depth;
+}
+
+inline void decrement_depth(multi_ex& ec_seq)
+{
+  if(vw_dbg::track_stack){
+    for (auto& ec : ec_seq)
+    {
+      --ec->stack_depth;
+    }
+  }
+}
+
```

Comment:
I am concerned about all of this debug functionality. It seems like there is state and logic around the debug logs. It looks like we are adding cognitive load to the developers with this. If we forget to increment/decrement what are the chances of us spending time debugging the debugging logs? One thing could be to leave the functionality here should we need it in the future, but remove it from the reduction code (I think someone else proposed this too).

Reply:
It's unfortunate that I called this VW_DBG instead of VW_TRACE.  

It's true that we are tracking the depth of stack for trace output.  This is isolated in the learner and is transparent to reduction writers.

At this point using VW_DBG in a reduction just works and produces output in the context of the reduction stack.  Trying to trace how prediction or learn works in a reduction stack is substantially harder without this.